[WIP] ComplexFloat class

cbm755 opened the pull request:
Eventually this will fix #12154.  Here for discussion about whether we want such a class...

### TODO list:
- [x] basic implementation
- [x] printing
    - [x] exact format of printing TBD.
    - [x] think about `1.2+3.4j` some more, 
    - [ ] code printing
- [ ] proper precision support
    - [x] ~~15 digits hardcoded in some places~~
    - [x] need tests for various precisions
    - [ ] need tests for mixed precision input for real and imag inputs
    - [x] dps and prec support, with dps default.
- [x] nan behaviour?
- [x] infinity behaviour?
- [x] Float greedily makes ComplexFloat by combining with `I` etc
    - [x] probably more cases to do
- [x] fix `conjugate()`.
- [x] fix a printing bug for ComplexFloat(...)*Symbol('x')
    `pprint(S(1+2j)*x)` looks like `1.0 + 2.0ⅈ⋅x`: it needs paren.
- [x] `ComplexFloat((1.1, 1.2))` should probably work
- [x] `ComplexFloat((<mpf>, <mpf>))` should probably work
- [x] `ComplexFloat(<mpc>)` should work
    - [x] tested?
- [x] `S(1+2j) % 2` causes recursion.
- [ ] @asmeuer said: "A real ComplexFloat should perhaps just return a Float." Perhaps so.  Maybe `._new` should do this but `__new__` should not.  (S.Zero already has special treatment in `._new`.
    - [ ] fft/ifft can pickup zero components of ComplexFloat, see discrete/transforms.py doctests
    - [ ] argument against doing this: signed zero is useful for branch cuts: `ComplexFloat(-1, -0.0)` vs `ComplexFloat(-1, 0.0)`.
- [ ] What should `Float(complex(1,2))` do?  What about `Float(ComplexFloat(1,2))`?  Should both return ComplexFloat?  Right now, they are errors.  I think this results in the test failures in:
   - [ ] geometry/test/test_ellipse.py
   - [ ] integrals/test/test_integrals.py 
- [x] sorting of lists of ComplexFloat and Floats is breaking a test in ` polys/tests/test_polyroots.py`.
- [ ] Some failing tests that I'm uncertain about:
    - [ ] core/tests/test_complex:  `4j*x == 4*x*I`
         For sure, `4j*x` should become `ComplexFloat(0,4)*x` is just a question as to whether that should be equal to the `4*x*I`.  For example, `ComplexFloat(0,4) == 4*I` is currently `True`.
    - [x] core/tests/test_function.py: `isinstance(sin((1.0 + 1.0*I)**10000 + 1), sin)`.
    - [x] sympy/core/tests/test_sympify.py: numpy thing, #13275 would fix this.
    - [x] sympy/solvers/tests/test_solvers.py: can reproduce on 1.1.1, filed #13298
    - [x] sympy/integrals/tests/test_meijerint.py: filed #14980, helps here.
    - [ ] TODO: some others to list...

### Possible future work
- [ ] storage: currently uses two Floats, but could use a `mpmath.mpc` which is already used for some arithmetic.
- [ ] `ComplexFloat("1.23 + 2.34j")`: should that work?  `complex('1+2j')` doesn't.
- [ ] Search code for `is_Float` and see what should be done for `ComplexFloat` (probably a rather big job)
- [ ] @asmeuer suggested a common subclass for Float and ComplexFloat.

#### Release notes

<!-- BEGIN RELEASE NOTES -->
* core
  * new ComplexFloat class improves floating point arithmetic
<!-- END RELEASE NOTES -->

cbm755 committed:
Add a ComplexFloat class

This will enscapulate `Float(a) + Float(b)*I`, allowing operators on
such.

cbm755 committed:
ComplexFloat: print negative imag part correctly

cbm755 committed:
ComplexFloat: parenthesize correctly by setting precedence

cbm755 committed:
ComplexFloat: fix conjugate, adjoint and Abs

cbm755 committed:
Mul: greedily combine Float, ComplexFloat and I

After the flattening, we may have a Float/ComplexFloat next to I.
These can be combined into a ComplexFloat.

cbm755 committed:
Add: greedily combine Float, ComplexFloat and I

After the flattening, we may have a Float/ComplexFloat in the head of the
list, followed by a Number*I.  These can make ComplexFloat.

cbm755 committed:
Basic: add "is_ComplexFloat" property

cbm755 committed:
ComplexFloat: _mpc_ property for converting to mpmath

cbm755 committed:
ComplexFloat cannot be converted to native float

cbm755 committed:
ComplexFloat: fix floor, ceiling and friends

cbm755 committed:
mpmathify: call ComplexFloat in the mpc case

I don't think this changes anything, but perhaps its a little faster
to call it directly?

cbm755 committed:
ComplexFloat: fix in Mul for when "I*Pi"

Maybe there are other cases and we cannot assume the "I" will move to
the front of the list.  But for now, special case for "Pi*I".

cbm755 committed:
ComplexFloat: switch printing to "a + b*I" and update tests

So many tests fail otherwise.  Also, start updating tests that now
need "0.0 + x.y*I" instead of "x.y*I".

cbm755 committed:
ComplexFloat: fixes and tests for nan and inf

cbm755 committed:
ComplexFloat: more consistent latex printing

cbm755 committed:
ComplexFloat: compare across type

Like current Float behaviour, allow ComplexFloat to equal things that
are not ComplexFloats.

asmeurer commented:
I've often thought it would be useful to have a complex type. Just doing it for floating point complex numbers seems like a good idea. That means that things like powers and products of non-float complex numbers can still be represented unevaluated.

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -1165,6 +1183,153 @@ def __format__(self, format_spec):
+
+    @property
+    def _prec(self):
+        return min(self.real._prec, self.imag._prec)
```
We are pretending that the ComplexFloat has a uniform precision, but storing the component precisions as they were? Would it be better to just do `prec = min(real._prec, imag._prec)` in the constructor? Actually, as it is now, it doesn't even look like you can have distinct precisions in the constructor.

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -1165,6 +1183,153 @@ def __format__(self, format_spec):
+    is_complex = True
+
+    def __new__(cls, real, imag=None, prec=None):
+        # TODO: do something with prec
```
Should also accept a string, Python builtin float/complex types, and tuple (like `_mpc_`).

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -1165,6 +1183,153 @@ def __format__(self, format_spec):
+    __slots__ = ['real', 'imag']
+
+    # each Float represents many real numbers,
+    # both erational and irrational.
```
typo

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -1165,6 +1183,153 @@ def __format__(self, format_spec):
+        return self.imag.is_zero
+
+    @property
+    def is_imaginary(self):
```
These should also use the `_eval_is_imaginary` form.

asmeurer replied:
Grep for `_eval_is_`. The methods define the assumptions (instead of using `@property`).

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -1165,6 +1183,153 @@ def __format__(self, format_spec):
+    def is_real(self):
+        # TODO: see is_rational above
+        #return False if self.imag.is_nonzero else None
+        return self.imag.is_zero
```
A real ComplexFloat should perhaps just return a `Float`.

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -1367,7 +1532,7 @@ def __add__(self, other):
                 #TODO: this can probably be optimized more
                 return Rational(self.p*other.q + self.q*other.p, self.q*other.q)
-            elif isinstance(other, Float):
+            elif isinstance(other, (Float, ComplexFloat)):
```
Should probably have a common superclass.

asmeurer commented:
Please test that everything works with higher than the default precision. There were too many instances in Float (some still not fixed!) of things assuming the default precision, or falling back to `float`, and I don't want to repeat the same mistakes.

asmeurer commented:
I'm unsure about the printing. Do you think it should print similar to `complex`, with parentheses no space around the operator, to emphasize that it is a single object? It could be confusing that there's a +, but it's not an Add. OTOH, I'm worried that printing it like that is bad for readability.

cbm755 commented:
I've made some progress on this offline---sorry for not pushing before you read through this :(  I will incorporate your suggestions.

re: printing: for now, I've gone back to "0.1 + 0.3*I" as that minimizes the changes needed to existing tests.  But I don't like it b/c it looks like an Add and a Mul.  I like your idea of no spaces: I'm thinking "0.1+0.3i" (without parenthesis) would be good eventually (or maybe `j`)

asmeurer commented:
j would make it valid Python.

cbm755 committed:
ComplexFloat: fix extract_minus_sign

cbm755 committed:
improve support for high-precision ComplexFloats

Add tests.

cbm755 committed:
ComplexFloat: add docstring

cbm755 committed:
ComplexFloat: fix up two failing tests

cbm755 committed:
Previously xfail test seems fixed

Seems to fail with python2 before ComplexFloat was added but now seems
fixed.

cbm755 committed:
ComplexFloat: str print similar to native complex type

So `0.1+0.2j`, no spaces around +/- and use `j` for unit imaginary.

cbm755 commented:
> These should also use the `_eval_is_imaginary` form.

@asmeurer What are those used for?  I grepped the code and don't see anything calling it...

> Should also accept a string, Python builtin float/complex types, and tuple (like _mpc_).

float/complex done

two strings done

added tuple/string to the todo list above

> We are pretending that the ComplexFloat has a uniform precision, but storing the component precisions as they were? Would it be better to just do prec = min(real._prec, imag._prec) in the constructor? Actually, as it is now, it doesn't even look like you can have distinct precisions in the constructor.

We now try to do something reasonable with precision when possible.  But if someone calls `ComplexFloat(Float(..., 10), Float(..., 32))` maybe we should respect that (maybe she wants to unpack it later for example).

cbm755 commented:
@asmeurer can you take another look?  Precision stuff is much improved and tested.  (CI is not going to pass, this is still just RFC).

I tried the `0.1-0.3j` printing style in the last commit: makes a lot of minor changes in `test_evalf`.  Some things are hard to visually parse:
````
`6.4471e-6-0.e-14j`
`1.23e-234+5.89e-123j`
````
(first is an actual example from `test_evalf`).  Might be better to put the spaces back in!

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -1173,6 +1204,321 @@ def __format__(self, format_spec):
+        if prec is None:
+            if isinstance(real, Float) and isinstance(imag, Float):
+                # no checks done; assume precisions are as desired by user
+                True
```
No-op should be `pass`.

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -1173,6 +1204,321 @@ def __format__(self, format_spec):
+    def _new(cls, _mpc_, _prec):
+        re, im = _mpc_
+        if (re, im) == (_mpf_zero, _mpf_zero):
+            return S.Zero  # XXX this is different from ComplexFloat()
```
I don't understand this comment.

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -1173,6 +1204,321 @@ def __format__(self, format_spec):
+
+    is_ComplexFloat = True
+
+    def __new__(cls, real, imag=None, prec=None):
```
Should allow `dps` and `prec` (see https://github.com/sympy/sympy/pull/12227).

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -1173,6 +1204,321 @@ def __format__(self, format_spec):
+        raise TypeError("can't convert ComplexFloat to float")
+
+    def __int__(self):
+        raise TypeError("can't convert ComplexFloat to int")
```
These should work if it is real (unless you are doing the real ComplexFloat returning Float thing?).

cbm755 replied:
re: real ComplexFloat returning Float: not sure yet, in arithmetic calculations maybe.

Anyway, in native Python `int(1+0j)` gives `TypeError: can't convert complex to int` and I think `__int__` is about conversion to native Python int.  Let me know if you think I should do this anyway.

asmeurer replied:
I didn't know that. I guess it is fine then.

asmeurer commented:
The printing doesn't bother me too much. That's also how Python's `complex` prints. For pretty printing there is `pprint` and `latex`, which should print ComplexFloat the same way as an `Add` with `I`.

cbm755 committed:
minor cleanup, reword comment

cbm755 committed:
ComplexFloat: use dps for decimal precision

Use `prec` for binary.  For now, this means we convert from decimal
to binary and then back to decimal before calling the Float ctor.
Assuming that will soon be fixed up for Float, leave some TODOs
highlighting the relevant lines.

cbm755 committed:
ComplexFloat: use _eval_is_real/complex instead of properties

cbm755 committed:
ComplexFloat: instantiate from strings and tuples

Not quite right because explicitly tests for tuple instead of duck
typing.  But its a start.  Add tests.

cbm755 committed:
ComplexFloat: implement extract_multplicatively

cbm755 committed:
ComplexFloat: partial revert, no longer needed for extract_minus

Revert part of commit 92034cf1b988626cdac84c3c7767039354fddf44 which
added a special case for ComplexFloat in `could_extract_minus_sign`.
Better fix is lower level so that `extract_multplicatively(-1)` works
and thus special case isn't needed anymore.  Keep the tests.

cbm755 committed:
ComplexFloat: fix pretty tests on Python 2

cbm755 committed:
ComplexFloat: some fixes for exprtool's Factor

I'm not very confident about these, leave a TODO.  Add a test
for a case that seemed to be untested.

cbm755 committed:
ComplexFloat: use "comp" for floating point tests

cbm755 committed:
ComplexFloat: support for mod

cbm755 committed:
re/im of ComplexFloat should return Float

Previously it didn't when ComplexFloat had zero real/imag part.

cbm755 committed:
ComplexFloat: update tests for formatting change

cbm755 committed:
ComplexFloat: exponent fixes and tests

Rational or Integer to power Float or ComplexFloat should evaluate.

cbm755 committed:
ComplexFloat: tests for Float powers that give Complex results

cbm755 committed:
Float: use ComplexFloat for power if result will be complex

cbm755 committed:
secondquant: do not assume all Numbers are real

Should be no harm in doing conjugate() even if they are real.  And
this fixes a bug with ComplexFloat.

cbm755 committed:
Float: extracting number from Float(0) should not give None

Add some tests for Float.extract_multplicatively (it didn't seem to
have any).

cbm755 committed:
extract_multiplicatively: more conservative for ComplexFloat

Behave more like "Add": generally just call extract_multiplicatively
on the real and imaginary parts.  In the special case when the real
part is zero, we could for example extract `I` or `constant*I`.
This means `sin((1+2j)*x)` doesn't generally become `sinh` on
instantiation.

cbm755 committed:
ComplexFloat: tighten up some tests

cbm755 committed:
ComplexFloat: more formatting fixes for tests

cbm755 committed:
test changed slightly because of ComplexFloat

Its only changed by roundoff and IMHO these tests are fragile, for
example depending on the order of floating point calculations.  Change
this one test to use `comp` within about 1.5*{machine epsilon} of the
hard-coded expected result.  Also, check its within a much weaker
tolerance of the specified exact solution.

cbm755 committed:
ComplexFloat: ensure Abs inherits precision

Fixes XFAIL test for exp/log.  Add another direct test.

cbm755 committed:
ComplexFloat: sort_key to sort these after other numbers

Increase the second entry of the sort key from 0 to 1: this seems to
put ComplexFloat's after all other Numbers when sorting.  At least one
test in exprtool relies on this.  Seems no more arbitrary then the old
behaviour which sorted by the real part compared to other real Numbers.

cbm755 committed:
minor fix so xfail tests fails for the right reasons

cbm755 committed:
ComplexFloat: comment about print-level formatting of Float

Behaviour of zeros in Float printing depends on whether the object is
by itself ("1.2500000000", "2.") or in an expression ("1.25", "2.0").
Note sure yet which we want for ComplexFloat but this commit adds some
tests that I worked out while understanding the difference.  Adds a
comment to the printing code about how the Float-like behaviour can
be achieved.

Also update the docstrings for expr.py for ComplexFloat printing;
these are subtle (its where I noticed this issue) and might need
updating if the printer behaviour changes.

cbm755 committed:
doctest updates for ComplexFloat printing

cbm755 committed:
Doc fix for ComplexFloat formatting

cbm755 committed:
Merge remote-tracking branch 'origin/master' into complexfloat_class_tmp_merge

cbm755 committed:
ComplexFloat: don't use deprecated Float(...,prec=)

Since the ComplexFloat work was started, Float's have been improved.
This simpliest various parts of the code too.

cbm755 committed:
ComplexFloat: duck type for .real and .imag instead of whitelist

In principle this would fix the np.complex256 error but it instead
exposes Issue #13275 :(

cbm755 committed:
Merge remote-tracking branch 'origin/master' into complexfloat_class

cbm755 committed:
Use ComplexFloat in Euler polynomial numerical evaluation

cbm755 committed:
Merge remote-tracking branch 'origin/master' into complexfloat_class

cbm755 committed:
Merge remote-tracking branch 'origin/master' into complexfloat_class

cbm755 committed:
test: slightly modifed version of this xfail test now passes

cbm755 committed:
test: should be fast but its ok if it evaluates

It will now that we have ComplexFloat.

cbm755 committed:
ComplexFloat: use is_Foo instead of isinstance

As per 835cb8259a77172c59d0c53c5a22f568efb8cb2c, this is faster but
does require that you know you have a sympy object.

cbm755 committed:
ComplexFloat: add TODO comments while we decide what we want

This test specifically states that `S(1.0)` and `S(1.0+0j)` do the same
thing and its not clear we want them too.  On the other hand, looking at
the original report, maybe its enough that both produce something floating
point, not necessarily the same thing...

cbm755 committed:
convolution: fix many accidental floats in tests

I don't think these are supposed to be python floats; many other
nearby quantities use `S` to ensure they are not floats.  So do more
of that.

asmeurer commented:
`numerics` was an old submodule that was removed a long time ago, which I included in the release notes bot accidentally (https://github.com/sympy/sympy-bot/pull/16). This should be listed under `core`, because that's where the code is.

cbm755 committed:
fft: doc that float ifft can add spurious imaginary components

Perhaps if ComplexFloat automatically reduces to Float, this could
be reverted.  (Although its still true that in floating point, imag
components of size machine epsilon can appear so the doc might be
useful anyway.)

cbm755 committed:
convolutions: adapt docs for ComplexFloat printing

cbm755 committed:
Revert "ComplexFloat: use is_Foo instead of isinstance"

This reverts commit 95c9f74d800fce8c85433c2b220897fa82141807.

Woah back, careful don't forget not everything in SymPy inherits from
Basic.  These kinds of improvments should be done carefully and not
as part of some huge PR...

cbm755 committed:
ComplexFloat: use is_Integer because we know its a Number

(and is_Integer is cheaper than isinstance)

cbm755 committed:
Use `isinstance()` instead of `is_Foo` in some cases

For consistency with surrounding code and safety for non-Basic
objects.

cbm755 committed:
updates for ComplexFloat printing changes

cbm755 committed:
ComplexFloat: fix pretty printing Mul

Not sure why this isn't based on precedence, so left a comment in the
code to look at again later.

cbm755 committed:
ComplexFloat: remove unused symbol from tests

cbm755 committed:
fix accidental floats in a test

I think these were supposed to be rationals

cbm755 committed:
codegen ast test: minor change for ComplexFloat

A ComplexFloat inherits the minimal precision of its real and imag
parts.  This is intentional.  The setup for this test assumed
otherwise.  Adjust the setup, but I believe the meaning of the test
is unchanged.

cbm755 committed:
factor_terms: don't assume `as_coeff_Mul()[0]` is real

As far as I can tell, this code really wants to know if -1 can be
extracted from each term.  Well, we don't need to check `< 0` to do
that; use `.extract_multiplicative(-1)` instead.  It should still be
cheap because `.as_coeff_Mul()` is supposed to be a cheap operation.

Before:
```
>>> %timeit factor_terms(-x - Catalan*I - 1)
483 µs ± 7.54 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

>>> %timeit factor_terms(-x - Catalan*I - 1)
474 µs ± 2.66 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

After:
```
>>> %timeit factor_terms(-x - Catalan*I - 1)
492 µs ± 8.87 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

>>> %timeit factor_terms(-x - Catalan*I - 1)
503 µs ± 12.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

cbm755 committed:
ComplexFloat: test that -1 can be factored out of `-x - S(1+2j)`

cbm755 committed:
polys: test fixes for ComplexFloat printing

cbm755 committed:
polys: replace accidental floats in tests

These are supposed to test against Rationals.

cbm755 committed:
quantum test: avoid floats in matrix

I don't think this test is really about floats, so create a Rational
Matrix instead.

cbm755 committed:
Merge remote-tracking branch 'origin/master' into complexfloat_class

czgdp1807 commented:
@cbm755 Please provide updates on this. Can this PR still be merged?

cbm755 commented:
@czgdp1807 it needs some work (and rebasing apparently) but in principle its still valid WIP.  Help of course welcome!

IIRC, I got hung up on bugs this uncovered.

I will see if I can find some time to rebase this at least.
